From 65fd1866fca679be8fb6559d7690c336ffce6e56 Mon Sep 17 00:00:00 2001
Message-Id: <65fd1866fca679be8fb6559d7690c336ffce6e56.1427265330.git.chang-joon.lee@intel.com>
In-Reply-To: <b4c66aa6a04d2be0fca9fa4448c7188a581a57a5.1427265330.git.chang-joon.lee@intel.com>
References: <b4c66aa6a04d2be0fca9fa4448c7188a581a57a5.1427265330.git.chang-joon.lee@intel.com>
From: sudheer gurram <sudheer.gurram@intel.com>
Date: Tue, 24 Mar 2015 15:25:34 +0530
Subject: [PATCH 22/22] Revert "SQUASHME! [VPG]: drm/i915: Remove unnecessary
 mutex around stats tracking."

This change was causing a regression in the race condition due to
non-synchronization of i915_gem_obj_insert_pid & i915_gem_obj_remove_pid

This patch reverts commit:
commit ac520ea5c73501aecddc60798dd9b68f76b2838f
Change-Id: I8d18fde4c41fde6f54226a073c1ddf97bc5ea682
Author: sudheer gurram <sudheer.gurram@intel.com>
Date:   Fri Feb 13 09:04:39 2015 +0530

    SQUASHME! [VPG]: drm/i915: Remove unnecessary mutex around stats tracking.

Issue: GMIN-7370
Change-Id: I046aa492fd428e5fa7b0978c6514e6bf2994544d
Signed-off-by: sudheer gurram <sudheer.gurram@intel.com>
---
 drivers/gpu/drm/i915/i915_gem_debug.c |   17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem_debug.c b/drivers/gpu/drm/i915/i915_gem_debug.c
index 988693f..80ac0d9 100644
--- a/drivers/gpu/drm/i915/i915_gem_debug.c
+++ b/drivers/gpu/drm/i915/i915_gem_debug.c
@@ -269,6 +269,10 @@ int i915_gem_obj_insert_pid(struct drm_i915_gem_object *obj)
 	if (!i915.memtrack_debug)
 		return 0;
 
+	ret = i915_mutex_lock_interruptible(obj->base.dev);
+	if (ret)
+		return ret;
+
 	list_for_each_entry(entry, &obj->pid_info, head) {
 		if (entry->tgid == current_tgid) {
 			entry->open_handle_count++;
@@ -280,14 +284,16 @@ int i915_gem_obj_insert_pid(struct drm_i915_gem_object *obj)
 		entry = kzalloc(sizeof(*entry), GFP_KERNEL);
 		if (entry == NULL) {
 			DRM_ERROR("alloc failed\n");
-			return -ENOMEM;
+			ret = -ENOMEM;
+			goto out;
 		}
 		entry->tgid = current_tgid;
 		entry->open_handle_count = 1;
 		INIT_LIST_HEAD(&entry->virt_addr_head);
 		list_add_tail(&entry->head, &obj->pid_info);
 	}
-
+out:
+	mutex_unlock(&obj->base.dev->struct_mutex);
 	return ret;
 }
 
@@ -296,11 +302,15 @@ void i915_gem_obj_remove_pid(struct drm_i915_gem_object *obj)
 	pid_t current_tgid = task_tgid_nr(current);
 	struct drm_i915_obj_pid_info *pid_entry, *pid_next;
 	struct drm_i915_obj_virt_addr *virt_entry, *virt_next;
-	int found = 0;
+	int ret, found = 0;
 
 	if (!i915.memtrack_debug)
 		return;
 
+	ret = i915_mutex_lock_interruptible(obj->base.dev);
+	if (ret)
+		return;
+
 	list_for_each_entry_safe(pid_entry, pid_next, &obj->pid_info, head) {
 		if (pid_entry->tgid == current_tgid) {
 			pid_entry->open_handle_count--;
@@ -319,6 +329,7 @@ void i915_gem_obj_remove_pid(struct drm_i915_gem_object *obj)
 			break;
 		}
 	}
+	mutex_unlock(&obj->base.dev->struct_mutex);
 
 	if (found == 0)
 		DRM_DEBUG("Couldn't find matching tgid %d for obj %p\n",
-- 
1.7.9.5

